File: frontend/src/engine/capabilities/addSparkline.ts
/**
 * addSparkline – Add sparkline mini-charts to a range of cells.
 *
 * Office.js notes:
 * - Sparklines require ExcelApi 1.9+.
 * - The correct API is worksheet.sparklineGroups.add(type, sourceData, locationRange)
 *   NOT range.group() — that method is for row/column outline grouping.
 * - SparklineType enum is not exported in all @types/office-js versions,
 *   so we pass the string values ("Line", "Column", "WinLoss") via cast.
 *
 * Typical use-case:
 *   dataRange:     "Sheet1!B2:M10"   (12 months × 9 products)
 *   locationRange: "Sheet1!N2:N10"   (one sparkline per product row in col N)
 */

import { CapabilityMeta, AddSparklineParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "addSparkline",
  description: "Add sparkline mini-charts to cells",
  mutates: false,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.9",
};

async function handler(
  context: Excel.RequestContext,
  params: AddSparklineParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { dataRange, locationRange, sparklineType = "line" } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would add ${sparklineType} sparklines to ${locationRange} from ${dataRange}`,
    };
  }

  options.onProgress?.(`Adding ${sparklineType} sparklines...`);

  const locationRng = resolveRange(context, locationRange);
  const dataRng    = resolveRange(context, dataRange);

  // Excel API string values for SparklineType (ExcelApi 1.9+).
  // We use strings instead of the Excel.SparklineType enum because the enum
  // is not exported in all versions of @types/office-js.
  const typeMap: Record<string, string> = {
    line:    "Line",
    column:  "Column",
    winLoss: "WinLoss",
  };
  const excelType = typeMap[sparklineType] ?? "Line";

  // sparklineGroups.add(type, sourceData, locationRange) — ExcelApi 1.9
  // Cast worksheet to any because SparklineGroupCollection typing may be absent.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const sparklineGroup: any = (locationRng.worksheet as any).sparklineGroups.add(excelType, dataRng, locationRng);

  // Apply optional color to the sparkline group
  if (params.color) {
    sparklineGroup.format.line.color = params.color;
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Added ${sparklineType} sparklines to ${locationRange} from ${dataRange}`,
    outputs: { range: locationRange },
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.9) ───────────────────────────────────
// Sparklines don't exist before 1.9. Best substitute: embed a small column
// chart per row positioned over each target cell. This works on Excel 2016+
// (charts were in 1.1) and visually mimics a sparkline — compact, inline, and
// driven by the same source data so it updates live.
async function fallback(
  context: Excel.RequestContext,
  params: AddSparklineParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { dataRange, locationRange, sparklineType = "line" } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would add ${locationRange.split(":").length > 1 ? "mini charts per row" : "a mini chart"} at ${locationRange} (sparklines unsupported; using embedded ${sparklineType === "column" ? "column" : "line"} charts).`,
    };
  }

  options.onProgress?.("Legacy-Excel mode: embedding mini charts instead of sparklines...");

  const dataRng = resolveRange(context, dataRange);
  const locRng = resolveRange(context, locationRange);
  dataRng.load(["rowCount", "columnCount", "address", "worksheet/name"]);
  locRng.load(["rowCount", "columnCount", "address", "left", "top", "width", "height", "worksheet/name"]);
  await context.sync();

  // One chart per location cell, driven by the corresponding row of source data.
  const rows = Math.min(dataRng.rowCount, locRng.rowCount);
  if (rows <= 0) {
    return { stepId: "", status: "error", message: "Data range and location range must each contain at least one row." };
  }

  const sheet = locRng.worksheet;
  const chartType: string = sparklineType === "column" ? "ColumnClustered" : "Line";

  for (let r = 0; r < rows; r++) {
    const rowData = dataRng.getRow(r);
    const anchor = locRng.getCell(r, 0);
    anchor.load(["left", "top", "width", "height"]);
    await context.sync();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const chart: any = (sheet as any).charts.add(chartType, rowData, "auto");
    // Size + position the chart to hug the anchor cell.
    chart.left = anchor.left;
    chart.top = anchor.top;
    chart.width = anchor.width;
    chart.height = anchor.height;
    // Strip non-essential chrome so it reads like a sparkline.
    try {
      chart.title.visible = false;
      chart.legend.visible = false;
      chart.axes.categoryAxis.visible = false;
      chart.axes.valueAxis.visible = false;
    } catch {
      /* some chart sub-properties may not be available on very old API sets */
    }
    if (params.color) {
      try {
        chart.series.getItemAt(0).format.line.color = params.color;
      } catch {
        /* color setting best-effort on legacy API */
      }
    }
    await context.sync();
  }

  return {
    stepId: "",
    status: "success",
    message:
      `Added ${rows} mini ${sparklineType === "column" ? "column" : "line"} charts to ${locationRange} ` +
      `(legacy-Excel fallback — native sparklines require ExcelApi 1.9+).`,
    outputs: { range: locationRange },
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/add_sparkline.py
"""
addSparkline — attach sparkline mini-charts to a location range.

Uses desktop Excel's VBA/COM SparklineGroups API (always present on Excel
2010+), so unlike the TS handler we don't need an ExcelApi-version fallback.
Type: 1=line, 2=column, 3=winloss.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


# Excel XlSparkType enum values.
_SPARK_TYPE = {
    "line": 1,
    "column": 2,
    "winloss": 3,
    "winLoss": 3,
}


def _hex_to_rgb_int(color: str) -> int:
    """#RRGGBB → Excel BGR long (R + G*256 + B*65536)."""
    h = color.lstrip("#")
    if len(h) != 6:
        raise ValueError(f"Invalid color {color!r}; expected #RRGGBB.")
    return int(h[0:2], 16) + int(h[2:4], 16) * 256 + int(h[4:6], 16) * 65536


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params.get("dataRange")
    location_range = params.get("locationRange")
    spark_type = (params.get("sparklineType") or "line")
    color = params.get("color")

    if not data_range or not location_range:
        return {
            "status": "error",
            "message": "addSparkline requires 'dataRange' and 'locationRange'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would add {spark_type} sparklines to {location_range} from {data_range}.",
        }

    try:
        data_rng = resolve_range(ctx.workbook_handle, data_range)
        loc_rng = resolve_range(ctx.workbook_handle, location_range)

        type_val = _SPARK_TYPE.get(spark_type, 1)

        # The SparklineGroups collection lives on the *location* sheet's COM
        # Range. API: location.SparklineGroups.Add(Type, SourceData) where
        # SourceData is the external address string (sheet-qualified).
        source_address = data_rng.sheet.name + "!" + data_rng.address.lstrip("=")
        # data_rng.address already includes the sheet prefix in many xlwings
        # versions — strip any duplicate and rebuild as "SheetName!A1:B2".
        cell_only = data_rng.address.split("!")[-1]
        source_address = f"{data_rng.sheet.name}!{cell_only}"

        group = loc_rng.api.SparklineGroups.Add(type_val, source_address)

        if color:
            try:
                group.SeriesColor.Color = _hex_to_rgb_int(color)
            except Exception:  # noqa: BLE001 — best-effort styling
                pass
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"Add sparkline failed: {exc}",
            "error": str(exc),
        }

    return {
        "status": "success",
        "message": f"Added {spark_type} sparklines to {loc_rng.address} from {data_rng.address}.",
        "outputs": {"range": loc_rng.address},
    }


registry.register("addSparkline", handler, mutates=False, affects_formatting=True)
